src/replay/scheduler.zig

```diff
@@ -0,0 +1,294 @@
+        switch (self.thread_pool.pollFallible()) {
+            .done => {
+                assert(self.batches_started == self.batches_finished);
+                if (self.failure) |f| {
+                    return .{ .err = f };
+                } else if (self.batches.items.len != self.batches_started) {
+                    if (try self.tryScheduleSome()) |err| {
+                        self.exit.store(true, .monotonic);
+                        self.failure = .{ .invalid_transaction = err };
+                    }
+                    return .pending;
+                } else {
+                    assert(self.batches.items.len == self.batches_finished);
+                    return .done;
+                }
+            },
+            .pending => return .pending,
+            .err => |err| return err,
+        }
+    }
+
+    fn tryScheduleSome(self: *TransactionScheduler) !?TransactionError {
+        while (self.batches.items.len > self.batches_started) {
+            const batch = self.batches.items[self.batches_started];
+            self.locks.lockStrict(self.allocator, batch.accounts) catch |e| {
+                switch (e) {
+                    error.LockFailed => if (self.batches_started == self.batches_finished) {
+                        return .AccountInUse;
+                    } else {
+                        break;
+                    },
+                    else => return e,
+                }
+            };
+            assert(try self.thread_pool.trySchedule(self.allocator, .{
+                .transactions = batch.transactions,
+                .results = &self.results,
+                .exit = &self.exit,
+            }));
+            self.batches_started += 1;
+        }
+        return null;
+    }
+};
+
+const ProcessBatchTask = struct {
+    transactions: []const ResolvedTransaction,
+    results: *Channel(?TransactionError),
+    exit: *std.atomic.Value(bool),
+
+    pub fn run(self: *ProcessBatchTask) !void {
+        const result = processBatch(self.transactions, self.exit);
+        if (result != null) self.exit.store(true, .monotonic);
+        try self.results.send(result);
+    }
+};
+
+test "TransactionScheduler: happy path" {
+    const allocator = std.testing.allocator;
+    const Transaction = sig.core.Transaction;
+    var rng = std.Random.DefaultPrng.init(123);
+
+    var thread_pool = ThreadPool.init(.{});
+    var scheduler = try TransactionScheduler.initCapacity(allocator, 10, &thread_pool);
+    defer scheduler.deinit();
+
+    const transactions = [_]Transaction{
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+        try Transaction.initRandom(allocator, rng.random()),
+    };
+    defer for (transactions) |tx| tx.deinit(allocator);
```

Comment:
This would leak a bunch if any of the transactions besides the first fail. I'd suggest using an arena instead, if the objective is terseness.
```suggestion
    var tx_arena_state: std.heap.ArenaAllocator = .init(allocator);
    defer tx_arena_state.deinit();
    const tx_arena = tx_arena_state.allocator();
    const transactions = [_]Transaction{
        try Transaction.initRandom(tx_arena, rng.random()),
        try Transaction.initRandom(tx_arena, rng.random()),
        try Transaction.initRandom(tx_arena, rng.random()),
        try Transaction.initRandom(tx_arena, rng.random()),
        try Transaction.initRandom(tx_arena, rng.random()),
        try Transaction.initRandom(tx_arena, rng.random()),
    };
```

Reply:
I think it is of consequence for there to be potential leaks, which in a failure state would lead to confusion seeing both the main error and the leaks. I don't think we should be lazy about memory allocation just because it's a unit test.